Framework/PythonInterface/plugins/algorithms/ANSTO/emu_common.py

```diff
@@ -0,0 +1,254 @@
+        self._table = table
+
+    @property
+    def frequency(self) -> float:
+        return self._frequency
+
+    @property
+    def amplitude(self) -> float:
+        return self._amplitude
+
+    @property
+    def speed(self) -> float:
+        return self._speed
+
+    @property
+    def phase(self) -> float:
+        return self._phase
+
+    def set_amplitude_speed(self, sample_file: str, ds_index: int = 0) -> None:
+        tags = [
+            HdfKey("dopp_amp", "/entry1/instrument/doppler/ctrl/amplitude", None),
+            HdfKey("dopp_vel", "/entry1/instrument/doppler/ctrl/velocity", None),
+        ]
+        values, _ = extract_hdf_params(sample_file, tags)
+        # mm to metres
+        self._amplitude = float(values["dopp_amp"][ds_index]) * 0.001
+        self._speed = float(values["dopp_vel"][ds_index])
+        self._complete_doppler_params()
+
+    def check_doppler_settings(self, ws_tag: str) -> None:
+        # confirm that the doppler amplitude and speed match that of the sample
+        mrun = mtd[ws_tag].getRun()
+        amp = mrun.getProperty("DopplerAmplitude").value[0]
+        spd = mrun.getProperty("DopplerVelocity").value[0]
+        if math.fabs(self._amplitude - amp) > 0.001 or math.fabs(self._speed - spd) > 0.1:
+            raise ValueError("Doppler parameter")
+
+    def _complete_doppler_params(self) -> None:
+        # performs a matrix inversion of the doppler table to create a linear regression
+        try:
+            rows = self._table.split(";")
+            table = []
+            for row in rows:
+                table.append([float(x) for x in row.split(",")])
+            table = np.array(table, dtype=float)
+        except ValueError:
+            raise RuntimeError("Invalid Doppler table entries in file.")
+
+        # the fitting works by finding the the closest entry in the table
+        # that matches the (speed, amp) if either phase or frequency is 'fixed'
+        # if the match is not close raise an exception
+        # note amp is in mm in table
+        speed, amp = self._speed, self._amplitude * 1000.0
+        min_ix = None
+        if self._frequency_option.lower() == "fixed" or self._phase_option.lower() == "fixed":
+            # find the closest match
+            distance = np.linalg.norm(table[:, :2] - [speed, amp], axis=1)
+            min_ix = np.argmin(distance)
+            if distance[min_ix] > 0.01:
+                raise RuntimeError("Cannot find suitable doppler entry for {}, {}.".format(speed, amp))
+
+        if min_ix and self._frequency_option.lower() == "fixed":
+            freq = table[min_ix, 2]
+        elif self._frequency_option.lower() == "auto":
+            freq = 0.5 * speed / (math.pi * self._amplitude)
+        else:
+            freq = float(self._frequency_option)
+
+        if min_ix and self._phase_option.lower() == "fixed":
+            phase = table[min_ix, 3]
+        else:
+            phase = float(self._phase_option)
```

Comment:
Same question here for the `self._phase_option`

Reply:
Similar to above.